src/liballoc/collections/btree/node.rs

```diff
@@ -1067,7 +1069,7 @@ impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::KV>
     /// - All the key/value pairs to the right of this handle are put into a newly
     ///   allocated node.
     pub fn split(mut self) -> (NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, K, V, Root<K, V>) {
-        debug_assert!(!self.node.is_shared_root());
+        assert!(!self.node.is_shared_root());
```

Comment:
AFAICT, all of the asserts for shared root in this block can be removed, as we have unique access to the node here (i.e. `marker::Mut<'a>`).

Reply:
But does node.rs impose or check any assumption on marker::Mut? Earlier, I [made a branch](https://github.com/rust-lang/rust/compare/master...ssomers:into_key_slice_mut_demo) that showed `into_key_slice_mut` doesn't work on the shared root, if we would start using it and subjected it to exotic alignment. Nothing stopped the code from making marker::Mut NodeRef to the shared root. In fact, that example contradicts your assertion that we can't have mutable reference to the shared root. `keys_mut(&mut self)` compiles and gets invoked fine on the shared root.

Reply:
Perhaps I misunderstood - my understanding is that the marker indicates unique access to the node. With the shared root, that's not possible. Note that we may be able to do it in code today, but it is always UB to have &mut to a static.

I am uncertain why miri does not catch us (or exactly what miri would be catching - I guess if we do just have a uniquely owned handle which internally has *mut to the shared root, then that would be fine. My recollection is that's not the case...

Reply:
I suppose the marker implies the wish for unique access (wish certainly not granted afer reborrow_mut and the many ptr::read's around), but what stops anyone from pretending to have unique access to the shared root? So it seems justified to me to have a (debug_)assert against that. Maybe not in the various functions on marker::Mut types, but in the function that creates a marker::Mut node reference.

Reply:
>  it is always UB to have &mut to a static

But does that imply that you can't `&mut` to a `NodeRef` instance whose field `node` is a NonNull pointer to a static? If so, I guess that's an area of improvement for Miri.

Reply:
No, it does not. I thought that `NodeRef` was not indirectly referencing a static, sorry. I guess the name "Ref" implies that.

Reply:
I think we can keep the debug asserts (or asserts, after this PR) as this PR makes them, I'll go through and resolve some comments.